### PR TITLE
ENH: add pytmc-debug GUI tool for inspecting pytmc processing

### PR DESCRIPTION
**Pull request by klauer**
image

Although this script has a dependency on `qtpy` (and therefore `PyQt5`/`Qt5`/etc), I think it can be considered optional. I had intended to make a command line-only alternative to the Qt version, but I'm not sure it's worthwhile.

### CODE REVIEW

**Comment by n-wbrown** [comment]
1. This is really cool! 
2. Do we want to add a second requirements file for this new tool (e.g. debug-gui-requirements.txt)?

**Comment by klauer** [comment]
1. Thanks, @n-wbrown! Think it's the best we can do as far as representing the chains/records/config information?

2. Hmm, maybe it could be added to `dev-requirements`?

**Comment by n-wbrown** [comment]
1. I think the only recommendations I could make are pretty minor. Maybe numbering the rows, labeling the individual windows, or indenting successively encapsulated layers? I find the interface is completely readable for a pytmc developer as is so I'm having a hard time recommending any real changes. If we intend to expand its usage so any TwinCAT developer can use it (which now that I'm thinking about it seems like it might be a great PLC debugging/development tool), then we might start making these small readability upgrades. If we start doing this, it could also be neat to feed the output of the linter into `pytmc_debug` and highlight where the issues originated. 

2. That seems appropriate.

File: pytmc/bin/pytmc.py [hunk 1]
```diff
@@ -14,19 +14,20 @@
 description = __doc__
 
 
-def make_db_text(tmc_file_obj, *, dbd_file=None, allow_errors=False,
-                 show_error_context=True):
+class LinterError(Exception):
+    ...
```
Comment (teddyrendahl) [inline pytmc/bin/pytmc.py]: Educate me. Why not `pass`?

File: pytmc/bin/pytmc_debug.py [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+    item_selected = Signal(object)
+
+    def __init__(self, tmc, dbd):
+        super().__init__()
+        self.tmc = tmc
+        self.chains = {}
+        self.records = {}
+
+        for record in tmc.all_RecordPackages:
+            record_text = record.render_record()
```
Comment (teddyrendahl) [inline pytmc/bin/pytmc_debug.py]: Think you might want to include this in the `try / except` block. `render_record` could raise some errors if we have an invalid `RecordPackage`

File: pytmc/bin/pytmc_debug.py [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+    dbd : DbdFile, optional
+        The dbd file to lint against
+    '''
+    app = QtWidgets.QApplication([])
+    interface = TmcSummary(tmc, dbd)
+    interface.show()
+    sys.exit(app.exec_())
+
+
+def main():
```
Comment (teddyrendahl) [inline pytmc/bin/pytmc_debug.py]: Might want to also add a hook for verbosity of `logging` stream?

File: pytmc/bin/pytmc.py (hunk 1 above)
Comment (klauer) [reply pytmc/bin/pytmc.py]: Either are equally valid, assuming 3.x+; so it's just a stylistic choice.

File: pytmc/bin/pytmc_debug.py (hunk 3 above)
Comment (klauer) [reply pytmc/bin/pytmc_debug.py]: I had it and thought it just added unnecessary spam during startup, when what I wanted to see was the GUI. If either you or @n-wbrown feel strongly about it, though, I'd be happy to add it back.

Comment (teddyrendahl) [reply pytmc/bin/pytmc_debug.py]: As in you had the `DEBUG` logging stream? You could just set it to `WARNING` unless otherwise specified?

Comment (klauer) [reply pytmc/bin/pytmc_debug.py]: Sorry - I just meant that this tool had the same `--log` flag as in the main `pytmc` utility. Added it back in 12a954c259648f884132f634a9433d2b4bf42517.

**Pull request merged**